### DISCUSSION TITLE
unique column label not produced in predictions - why?

### DISCUSSION THREAD

**Discussion by jkginfinite** [moj-analytical-services/splink#1505] [Q&A]
Hello,
I have been able to make predictions of duplicates on 10million rows.
How ever for some reason the unique column label (renamed to master record id, which is specified in the settings) is not produced in the predicted rows;

Why is this and how can this be changed? i need the master record id in my predictions

**Comment by RobinL** [comment]
It seems to work for me - don't suppose you're able to provide a reprex of the bug you're seeing?

```python
import pandas as pd
from splink.duckdb.linker import DuckDBLinker
from splink.duckdb.comparison_library import exact_match

data = [
    {"master_id": 1, "first_name": "John", "surname": "Smith"},
    {"master_id": 2, "first_name": "John", "surname": "Smith"},
    {"master_id": 3, "first_name": "Lucy", "surname": "Jones"},
]

settings = {
    "unique_id_column_name": "master_id",
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [],
    "comparisons": [
        exact_match("first_name"),
        exact_match("surname"),
    ],
}

linker = DuckDBLinker(pd.DataFrame(data), settings)

df_predict = linker.predict()
print(df_predict.as_pandas_dataframe().to_markdown())
```

Results in 

|   match_weight |   match_probability |   master_id_l |   master_id_r | first_name_l   | first_name_r   |   gamma_first_name | surname_l   | surname_r   |   gamma_surname |
|---------------:|--------------------:|--------------:|--------------:|:---------------|:---------------|-------------------:|:------------|:------------|----------------:|
|        6.71243 |          0.990554   |             1 |             2 | John           | John           |                  1 | Smith       | Smith       |               1 |
|      -23.2876  |          9.7666e-08 |             1 |             3 | John           | Lucy           |                  0 | Smith       | Jones       |               0 |
|      -23.2876  |          9.7666e-08 |             2 |             3 | John           | Lucy           |                  0 | Smith       | Jones       |               0 |

Similarly in spark:

```python
from splink.spark.linker import SparkLinker
from splink.spark.comparison_library import exact_match

from pyspark import SparkContext, SparkConf
from pyspark.sql import SparkSession
from pyspark.sql import Row


conf = SparkConf()
# This parallelism setting is only suitable for a small toy example
conf.set("spark.driver.memory", "12g")
conf.set("spark.default.parallelism", "16")

sc = SparkContext.getOrCreate(conf=conf)
spark = SparkSession(sc)
spark.sparkContext.setCheckpointDir("./tmp_checkpoints")

data = [
    {"master_id": 1, "first_name": "John", "surname": "Smith"},
    {"master_id": 2, "first_name": "John", "surname": "Smith"},
    {"master_id": 3, "first_name": "Lucy", "surname": "Jones"},
]

df = spark.createDataFrame(Row(**x) for x in data)

settings = {
    "unique_id_column_name": "master_id",
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [],
    "comparisons": [
        exact_match("first_name"),
        exact_match("surname"),
    ],
}

linker = SparkLinker(df, settings)

df_predict = linker.predict()
print(df_predict.as_pandas_dataframe().to_markdown(index=False))

```

results in 

|   match_weight |   match_probability |   master_id_l |   master_id_r | first_name_l   | first_name_r   |   gamma_first_name | surname_l   | surname_r   |   gamma_surname |
|---------------:|--------------------:|--------------:|--------------:|:---------------|:---------------|-------------------:|:------------|:------------|----------------:|
|        6.71243 |          0.990554   |             1 |             2 | John           | John           |                  1 | Smith       | Smith       |               1 |
|      -23.2876  |          9.7666e-08 |             2 |             3 | John           | Lucy           |                  0 | Smith       | Jones       |               0 |
|      -23.2876  |          9.7666e-08 |             1 |             3 | John           | Lucy           |                  0 | Smith       | Jones       |               0 |

**Reply by Stephaniav1901** [reply]
I did just as you said, but for some weird reason, the mater_id_l and mater_id_r don't match with the initial mater_id column. So I can't see which real rows are being compared by the end. Why is this happening?

**Reply by RobinL** [reply]
That could be the case I'd you're using

pyspark.sql.functions.monotonically_increasing_id()
 https://spark.apache.org/docs/3.1.3/api/python/reference/api/pyspark.sql.functions.monotonically_increasing_id.html#:~:text=A%20column%20that%20generates%20monotonically,in%20the%20lower%2033%20bits

I recommend avoiding this function if you're using it.

If you're not using it, are you able to provide a reprex of the problem?
https://data.library.virginia.edu/ask-better-code-questions-and-get-better-answers-with-reprex/

**Reply by Stephaniav1901** [reply]
Actually I tried using "pyspark.sql.functions.monotonically_increasing_id()" then I tried w = Window.partitionBy(lit(1)).orderBy(lit(1))
df_final =  df_final.withColumn("unique_id", row_number().over(w)).

Let's say that here df_final it's just like your's 'data' df but without this previous index column and it's already a Sparks dataframe.

`from splink.spark.linker import SparkLinker
from splink.databricks.enable_splink import enable_splink
import splink.spark.spark_comparison_library as cl

w = Window.partitionBy(lit(1)).orderBy(lit(1))
df_final =  df_final.withColumn("unique_id", row_number().over(w))

settings = {
    "unique_id_column_name": unique_id",
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [],
    "comparisons": [
        exact_match("first_name"),
        exact_match("surname"),
    ],
}

linker = SparkLinker(df_final,settings,validate_settings=True,spark=spark)

predictions = linker.predict()
pdf_predictions = predictions.as_pandas_dataframe()
spark.createDataFrame(pdf_predictions).display()
`
So by the end my prediction look like this:
<img width="471" alt="image" src="https://github.com/moj-analytical-services/splink/assets/49068355/98de51d2-5e7b-4628-8518-19ebbbde024f">
but at df_final my 'unique_id' 1144 and 538653 don't match with the names that are being compared on predicitons. 
So what might be happening?

**Reply by RobinL** [reply]
If you want to reliably assign a unique id to a Spark dataframe the most reliable way to do it is assign the unique id and then save the data out to parquet.  Then read it back in and proceed with using Splink

Other methodologies are prone to annoying gotyas due to Spark's distributed architecture

**Reply by Stephaniav1901** [reply]
It worked!!! Finally!!! Thank you so much!!! And the predictions are great!! This library it's excellent!
